File: solutions/go/protein-translation/1/protein_translation.go
package protein

import "errors"

const (
	Stop          = "STOP"
	Methionine    = "Methionine"
	Phenylalanine = "Phenylalanine"
	Leucine       = "Leucine"
	Serine        = "Serine"
	Tyrosine      = "Tyrosine"
	Cysteine      = "Cysteine"
	Tryptophan    = "Tryptophan"
)

var (
	ErrStop        error = errors.New("error stop")
	ErrInvalidBase error = errors.New("error invalid base")
	// CodonToProtein dictionary
	CodonToProtein = map[string]string{
		"AUG": Methionine,
		"UUU": Phenylalanine,
		"UUC": Phenylalanine,
		"UUA": Leucine,
		"UUG": Leucine,
		"UCU": Serine,
		"UCC": Serine,
		"UCA": Serine,
		"UCG": Serine,
		"UAU": Tyrosine,
		"UAC": Tyrosine,
		"UGU": Cysteine,
		"UGC": Cysteine,
		"UGG": Tryptophan,
		"UAA": Stop,
		"UAG": Stop,
		"UGA": Stop,
	}
)

// FromCodon transform codon to Protein value
func FromCodon(s string) (string, error) {
	value, found := CodonToProtein[s]
	if !found {
		return "", ErrInvalidBase
	}
	if value == Stop {
		return "", ErrStop
	}
	return value, nil
}

func FromRNA(s string) ([]string, error) {
	var result []string
	codons := splitByN(s, 3)
	for _, codon := range codons {
		protein, err := FromCodon(codon)
		switch err {
		case ErrStop:
			return result, nil
		case ErrInvalidBase:
			return result, err
		default:
			result = append(result, protein)
		}
	}
	return result, nil
}
func splitByN(s string, n int) []string {
	var codons []string
	length := len(s)
	for i := 0; i < length; i++ {
		if i%3 == 0 {
			if i+3 < length {
				codons = append(codons, s[i:i+3])
			} else {
				codons = append(codons, s[i:])
			}
		}
	}
	return codons
}

File: solutions/javascript/protein-translation/1/protein-translation.js
/**
 * A map of codons to their corresponding proteins.
 * @constant {Object<string, string>}
 */
const ACID_PROTEIN_MAP = {
  AUG: 'Methionine',
  UUU: 'Phenylalanine',
  UUC: 'Phenylalanine',
  UUA: 'Leucine',
  UUG: 'Leucine',
  UCU: 'Serine',
  UCC: 'Serine',
  UCA: 'Serine',
  UCG: 'Serine',
  UAU: 'Tyrosine',
  UAC: 'Tyrosine',
  UGU: 'Cysteine',
  UGC: 'Cysteine',
  UGG: 'Tryptophan',
  UAA: 'STOP',
  UAG: 'STOP',
  UGA: 'STOP',
};

/**
 * Gets the protein corresponding to a given codon.
 * @param {string} codon - The RNA codon.
 * @returns {string} The corresponding protein or 'INVALID' if the codon is not recognized.
 */
const getProtein = (codon) => ACID_PROTEIN_MAP[codon] || 'INVALID';

/**
 * Translates an RNA strand into a list of proteins.
 * @param {string} rnaStrand - The RNA strand to translate.
 * @returns {string[]} An array of proteins translated from the RNA strand.
 * @throws {Error} If an invalid codon is encountered in the RNA strand.
 */
export const translate = (rnaStrand) => {
  const proteins = [];

  if (rnaStrand) {
    for (let i = 0; i < rnaStrand.length; i += 3) {
      const protein = getProtein(rnaStrand.substring(i, i + 3));

      if (protein) {
        if (protein === 'STOP') {
          break;
        }

        if (protein === 'INVALID') {
          throw new Error('Invalid codon');
        }

        proteins.push(protein);
      }
    }
  }

  return proteins;
};
